keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,107 @@
+
+### Goals
+
+* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
+* Allow users to install a specific version of an application and/or a specific revision of an operator
+* Reject packages that don't have a matching KUDO API version
+
+### Non-Goals
+
+* Handle operator dependencies on specific Kubernetes versions
+
+## Proposal
+
+A `operator.yaml` file already provides versions for all components describing an operator:
+  * The dependency on KUDO is provided by the `apiVersion` and `kudoVersion` fields
+  * The bundled application version is provided by the `appVersion` field
+  * The version of the operator is provided by the `version` field
+  * The version of Kubernetes is provided by the `kubernetesVersion` field
+
+From these versions, currently only the `version` field can be used when installing packages with `kubectl kudo install`. Furthermore, a package repository index only contains `appVersion` and `version` fields in its entries. This limits the possible installation scenarios.
+At least the following installation scenarios should be supported:
+  1. A package repository could contain operators for older `apiVersion` and than the currently released KUDO. A user should be able to install these operators with older versions of KUDO.
+  2. A user should be able to install a specific `appVersion` of an operator
+
+### Support multiple API versions in the package repository
+
+Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
+
+For this, the following changes have to be made in KUDO:
+
+* Add `apiVersion` to the package index by updating the `Metadata` struct in `pkg/kudoctl/util/repo` and filling these values in `pkg/kudoctl/util/repo.ToPackageVersion`. A `index.yaml` entry should then look like this:
+
+```
+  kafka:
+  - appVersion: 2.3.0
+    digest: e80c7b783d327190d489159e89e0a005a6a8b00610bdb7e8b1bea73c49bf485a
+    maintainers:
+    - email: zmalikshxil@gmail.com
+      name: Zain Malik
+    name: kafka
+    urls:
+    - https://kudo-repository.storage.googleapis.com/kafka-0.2.0.tgz
+    version: 0.2.0
```

Comment:
we already mention that this version follows `semver`. Should we also say that it is immutable? 
My reasoning behind this is to clarify that if operator developer introduces a feature `advanced healthcheck` 
and its included in the release notes for `0.2.1`
And if user installs version `0.2.1` independent of the app version, should they expect the `advanced healthcheck` feature to be present there?

Reply:
> And if user installs version 0.2.1 independent of the app version, should they expect the advanced healthcheck feature to be present there?

Based on what was discussed last week, the operator version is reset on new minor/major app version releases, so a `0.2.1` operator version is going to mean different things on different app versions.

I added some details in this regard to the KUDO C* development documentation, right below the [example timeline table](https://github.com/mesosphere/kudo-cassandra-operator/blob/master/DEVELOPMENT.md#example-hypothetical-timeline-for-releases): 
> It's important to note that operator versions for different app versions are unrelated. e.g., in the example above both 3.11.4-0.1.0 and 4.0.0-0.1.0 have 0.1.0 as the operator version, but wouldn't necessarily share any commonality with regards to the operator itself. The operator version progression is only meaningful within an app version's major.minor family, i.e. 3.11.x and 4.0.x.

Reply:
but we should add to this doc then.
I see a confused support process in that approach, where we cannot say for sure which operator version brings feature A and which version is brining feature B.

if version `0.2.0` means different things based on the app version, it's not an operator version anymore. 
Because to represent an operator version we should guarantee the set of features its coming with. 

when installing kafka `2.3.0` we have that set of guarantees, which features will be in.
and installing operator version `0.2.0` won't give us the same set of guarantees as it will depend on which app version we are installing.

Reply:
> but we should add to this doc then

I agree. I had a chat with @nfnt and we'll bring some of the outcomes from last week's discussions and make them explicit in the KEP.

Reply:
but I still don't see them the most optimized way to do releases and introduce features. 
All this is to just introduce base-tech version? If that's the only benefit I still vote for 1 single semver that is immutable. And drop the base-tech versions at all from the versions. If we are having hard time to predict where a `feature A` belongs it's not going to help us but bring more complexity for end-users and support organizations.

Reply:
I'll add the initial pros/cons of each approach in the KEP so that we can continue the discussion. Although, I feel that most folks are leaning towards composite versions given the pros/cons discussed last week.

Reply:
I thought we still have to talk about that. But more I think about it more I can see that it will bring us issues related to unclear upgrade paths and operatorversion `semver` will render to be inefficient as it provides no meaning, where base tech provides some context and understanding. 

And we really need to understand the need for this. If the only need is to see visually the base-tech version in the release tag. I don't see it as enough reason to include it. We can introduce other mechanisms as we talked above to know the base tech version. And follow true semver for the operator releases.

Reply:
From the perspective of KUDO, we should avoid adding additional rules apart from semver on how this `version` field should look like. Resetting `version` on every `appVersion` change should work as well as having a `version` that is immutable to all `appVersion` changes.

Reply:
Though requiring `version` to be immutable (and thus independent from `appVersion`) has it's benefits as well, as it'll allow us to fully tread `appVersion` as a free-form string that doesn't have any implications/precedence on other versions. It would also allow for a list of `appVersions` as discussed above as we don't have to make _any_ assumptions here, only add logic to allow users to specify a `appVersion` and filter using it.

Reply:
right! we should be thinking how we will know and install the latest package when packages will look like:
`sarge-1.0.0`
`etch-1.0.0` 
`strech-1.0.0`
if that means more metadata field holding the order of the packages we should add it here. 

Having immutable versions would solve this out of the box

Reply:
With regards to resetting the `version` if the `appVersion` changes: I don't think this is a requirement, and left to the operator-developer. I think it would be a feasable approach to continue the versioning scheme:

````
kafka-2.3.0_1.0.0
kafka-2.3.0_1.1.0
kafka-2.3.1_1.1.0

kafka-2.3.0_1.1.1
kafka-2.3.1_1.1.1
```